### PR TITLE
[rush-lib] Fix `PreferredVersions` being ignored when projects have overlapping dependencies

### LINKED ISSUE #3205: [rush] 'rush update' ignores preferredVersions
## Summary

Preferred version is not respected. 
When a package specifies a range, 
the latest version is installed instead of the preferred version 
(which is lower but within the range).

Example from the repro repo:
Project "a" depends on `core-js@^3.0.0`
preferredVersion in common-versions.json is `"core-js": "3.6.5"`
installed version is `core-js: 3.21.0`

## Repro steps

clone https://github.com/ddilcheva/rush-common-deps-repro
```
rush install
> succeeds

rush check

> logs:
Node.js version is 14.18.2 (LTS)


Starting "rush check"

core-js
  ^3.0.0
   - a
  3.6.5
   - preferred versions from common-versions.json

Found 1 mis-matching dependencies!

rush update
> no updates 
```

  **Expected result:** `rush update` pins core-js@3.6.5 in common/rush/config/pnpm-lock.yaml

  **Actual result:** `core-js: 3.21.0` is installed

## Details

Specifying preferredVersion *first*, 
then adding dependency to the package (`rush add -p core-js@^3.0.0`)
works correctly and installs v3.6.5

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.61.4 |
| `rushVersion` from rush.json? | 5.61.4 |
| `useWorkspaces` from rush.json? | no |
| Operating system? | Mac |
| Would you consider contributing a PR? | No|
| Node.js version (`node -v`)? | v14.18.2 |

### PR DESCRIPTION
**Pull request by ace-n-msft**
## Summary

Fixes #3205

This PR changes `rush-lib` to prefer the most-restrictive SemVer range available across both `preferredVersions` and Rush projects.

### Motivation

The repos my team works with specify some dependencies on a repo-wide basis. We want these dependencies to be included in each project **and** the ability to bump repo-wide dependency pins with a single file change.

> Setting `ensureConsistentVersions = true` isn't sufficient, as (AFAICT) that would still require updating each `package.json` file to bump its dependency pins.

We tried to solve this by setting the individual project `package.json` files to SemVer ranges (e.g. `8.x` or `*`) and providing pinned dependency versions in the `preferredVersions` section of `common-versions.json`.

In the course of testing this approach, we ran into #3205 - hence this PR.

### Alternate approach: test isolation
Right now, this PR applies `common/config/rush/*` to **all** the tests in the `libraries/rush-lib/src/logic/pnpm/test`. This isn't causing any issues right now, but it may need to be refactored if other tests start depending on the `common/config/rush` directory.

If it's worth refactoring this, please let me know.
 
### Alternate approach: conditional `Map` overwriting
For this PR, I needed a mechanism to _conditionally_ overwrite the keys in a `Map`.

I was considering modifying the `MapExtensions.mergeFromMap()` function to accept a third parameter `shouldOverwrite` that would allow an end-user to control when existing keys should be overwritten.

Instead, I chose to do [a "manual" merge](https://github.com/microsoft/rushstack/pull/4835/commits/0a381c5985da213becee647b33c317a5a6271b8b#diff-cea1b277db05dae278b1d1872982fa2de3a47a57f3d846fcaeba95047f886512R99-R104) in `PnpmfileConfiguration.ts` using a `forEach()` loop.

### Possible breaking change: conflicting dependency versions across projects

According to the **Implicitly Preferred Versions** section in the [Rush docs](https://rushjs.io/pages/advanced/preferred_versions/#implicitly-preferred-versions), Rush has no specific convention for resolving dependency conflicts between projects.

As such, I wasn't sure if I should add automated tests for that case. **Please let me know if I should change this**, and what those test cases might look like.

Here are a few (additional) test cases that I thought might be worth implementing:

> **Repo structure**
> ```
> # project-1/package.json
> "dependencies": {
>     "library-a": "^1.0.0"
> }
> ```
>
> ```
> # project-2/package.json
> "dependencies": {
>     "library-a": "^1.2.0"
> }
> ```
>
> **Possible test case 1: should we check that this resolves to _only_ 1.2.0**
> ```
> # common-versions.json
> "preferredVersions": {
>     "library-a": "1.2.0"
> }
>
> # expected version set
> { "library-a": "1.2.0" }
> ```
>
> **Possible test case 2: should we do anything specific here?**
> ```
> # common-versions.json
> "preferredVersions": {
>     "library-a": "1.1.0"
> }
>
> # expected version set
> { "library-a": ["1.0.0"?, "1.1.0"?, "1.2.0"] }

## How it was tested

I've added automated tests for the common cases I could think of and documented some potential edge cases above.

(If those edge cases are things we should test for, I'm 100% onboard with adding automated tests for them as well.)

## Impacted documentation

We should consider adding a note to [this page](https://github.com/microsoft/rushstack-websites/blob/main/websites/rushjs.io/docs/pages/advanced/preferred_versions.md) explaining that Rush (now) prefers the _most specific_ version of a dependency available.

### CODE REVIEW

File: libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts [hunk 1]
```diff
@@ -91,11 +92,16 @@ export class PnpmfileConfiguration {
...
       const commonVersionsConfiguration: CommonVersionsConfiguration = subspace.getCommonVersions();
       const preferredVersions: Map<string, string> = new Map();
-      MapExtensions.mergeFromMap(preferredVersions, commonVersionsConfiguration.getAllPreferredVersions());
       MapExtensions.mergeFromMap(
         preferredVersions,
         rushConfiguration.getImplicitlyPreferredVersions(subspace)
       );
+      commonVersionsConfiguration.getAllPreferredVersions().forEach((version, name) => {
+        // Use the most restrictive version range available
+        if (!preferredVersions.has(name) || semver.subset(version, preferredVersions.get(name)!)) {
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: What happens when two ranges partially overlap but are not a strict subset are provided? Ex:

@my/package package.json:
```json
{
  ...
  "dependencies": {
    "@my/dependency": "^3.1.0"
  }
}
```

common-versions.json
```json
{
  ...
  "preferredVersions": {
    "@my/dependency": ">=3.0.0 <3.2.0"
  }
}
```

Ideally this would result in a version range of `>=3.1.0 <3.2.0`, but I believe your current change would cause it to remain as `^3.1.0` because `^3.1.0` is not a subset of `>=3.0.0 <3.2.0`

**Comment by D4N14L** [comment]
Regarding the test cases you mentioned:

`Possible test case 1`: Yes, this seems like a good test case to implement. Seems like exactly the scenario you're trying to solve here.

`Possible test case 2`: Assuming that `ensureConsistentVersions = false`, then I would imagine that the expected version sets would be `{ "library-a": ["1.1.0", "^1.2.0"] }`, since the latter range lies outside of the preferred version of `1.1.0`

File: libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts (hunk 1 above)
Comment (ace-n-msft) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: > You could find the intersection of the largest subset of those ranges to reduce the number of different-versioned packages consumed, however this would be rather complicated to determine

I thought about this too - `semver` itself doesn't do this, [but `semver-intersect` does](https://www.npmjs.com/package/semver-intersect). It seems to work quite well when tested locally!

The other option would be to warn the user whenever `preferredVersions` and `package.json` conflict. Perhaps via a message like the one below in Rush itself, and a more detailed explanation in the official docs?

```
@my/package (my_package/package.json): @my/dependency's version ranges in `my_package/package.json` and `preferredVersions` are ambiguous; `preferredVersions` will be ignored for this project.
```

**Comment by ace-n-msft** [comment]
@D4N14L thanks for the review!

Some news regarding the extra test cases - looks like there was a misunderstanding on my part. **It turns out** that the `getImplicitlyPreferredVersions()` function only returns versions for packages that have a _single SemVer range within the entire repository_. (So, by definition, packages with multiple unique SemVer ranges - _even overlapping ones!_ - should be excluded.)

> The `getImplicitlyPreferredVersions()` function ultimately calls out to the `DependencyAnalyzer`, which uses [this logic](https://github.com/microsoft/rushstack/blob/0e14ced009489874137403403a896372b50dedef/libraries/rush-lib/src/logic/DependencyAnalyzer.ts#L131-L164) to identify implicitly-preferred versions.

If we wanted to support overlapping SemVer ranges with `implicitlyPreferredVersions` (the behavior I discussed in the additional test cases), we could do that by adding a step to the `DependencyAnalyzer` [here](https://github.com/microsoft/rushstack/blob/0e14ced009489874137403403a896372b50dedef/libraries/rush-lib/src/logic/DependencyAnalyzer.ts#L147-L156).

-------

Based on #1466, it seems like we **do** want support for overlapping SemVer ranges. The changes to support that would look something like this:

```
--- a/libraries/rush-lib/src/logic/DependencyAnalyzer.ts
+++ b/libraries/rush-lib/src/logic/DependencyAnalyzer.ts
@@ -147,10 +147,15 @@ export class DependencyAnalyzer {
             if (implicitlyPreferredVersion === undefined) {
               // There isn't a candidate for an implicitly preferred version yet. Set this value as a candidate.
               implicitlyPreferredVersion = version;
+            } else if (semver.subset(version, implicitlyPreferredVersion)) {
+              // The previous candidate is more restrictive than this value. Set the candidate to this value.
+              implicitlyPreferredVersion = version;
+            } else if (semver.subset(implicitlyPreferredVersion, version)) {
+              // This value is less restrictive than the previous candidate. Do nothing.
             } else {
-              // There was already another version that was a candidate. Clear that out and break.
+              // The current value conflicts with the candidate. Clear the candidate out and break.
               // This dependency does not have an implicitly preferred version because there are at least
-              // two candidates.
+              // two conflicting candidates.
               implicitlyPreferredVersion = undefined;
               break;
             }
```

Maintainers - LMK if that change is something you'd want added to this PR. 🙂

**Comment by dmichon-msft** [comment]
> Maintainers - LMK if that change is something you'd want added to this PR. 🙂
I'm a fan of version deduplication via constraints; I'm also a fan of having the preferredVersion override allowedAlternativeVersions if the preferredVersion is a subset, but that might be a breaking change.

For a static validator a while back I wrote a semver-compatible range-intersection library, though I didn't have occasion to open-source it. If we end up needing that functionality I can contribute it to rushstack.

**Comment by ace-n-msft** [comment]
Circling back on this - @D4N14L  I met with @octogonz and @iclanton today; they're broadly on board with this PR save for one question.

When `preferredVersions` and `implicitPreferredVersions` have overlaps, should we:
1. have `implicitPreferredVersions` narrow down `preferredVersions`' constraints, _or_
2. have `preferredVersions` overwrite any `implicitlyPreferredVersions` entirely?

In practice, I feel like this is a tradeoff between **package freshness** and **disk space**.
- Narrowing down `preferredVersions` with the `implicitlyPreferredVersions` ranges would allow us to reuse package versions that are already used elsewhere within the repo.
- Overwriting `implicitlyPreferredVersions` would (in theory) allow Rush and the underlying package manager to the newest-possible package within the `preferredVersions` range.

Thoughts?

**Comment by D4N14L** [comment]
@ace-n-msft Hmm. So let me write a few things out to help frame this discussion (and to be honest, for myself to re-acclimate to the question at hand 😊):
- The purpose of preferred versions is meant to be used to enforce that the same version of a package is used across the repo, which is specifically important when bundling packages and ensuring you don't pull multiple versions of a package down when (for example) browsing to a website
  - `allowedAlternateVersions` is used specifically to break this enforcement for a specific version range. It is intended to be an "exception" to the rule
- Implicitly preferred versions are intended to be used for the same purpose, but once a second version is specified in the repo, they can no longer be "implicitly preferred"
- Both of these are used to attempt to constrain the versions of packages consumed by internal and external (non-repo) dependencies. If the preferred version is fully within the range specified by the external dependency, the preferred version range will be substituted. Otherwise it will be left as-is. This can be seen here: https://github.com/microsoft/rushstack/blob/7e16111340f7fb1b48ed8c40b5c193efab5134a0/libraries/rush-lib/src/logic/pnpm/PnpmfileShim.ts#L82

Referring to the above, the `implicitlyPreferredVersion` for a given package is either the existing package version across the repo or `undefined`. Given this, we should probably follow the existing behaviour used when replacing external package dependency versions (again, seen here: https://github.com/microsoft/rushstack/blob/7e16111340f7fb1b48ed8c40b5c193efab5134a0/libraries/rush-lib/src/logic/pnpm/PnpmfileShim.ts#L82 ) and override the `implicitlyPreferredVersion` with the explicit `preferredVersion` if and only if the explicit `preferredVersion` is entirely a subset of the `implicitlyPreferredVersion`. Essentially, this means that:
- `implicitlyPreferredVersions` are a no-op to repo packages, but continue to impact external packages
- Explicit `preferredVersions` now impact the repo, regardless of the state of `ensureConsistentVersions`
- The explicit version takes the priority in cases where there is both an explicit and an implicit version, which _feels_ correct
- We are performing the same expected logic on repo dependencies as we do on transitive dependencies

One side-effect of this is that `ensureConsistentVersions` basically becomes a linting tool to enforce that the specified version of a package in "package.json" is the version you get on install. One big note about this is that it means that during `rush publish`, the version of the dependencies in the packaged/published tarball will be updated to the version in the `common-versions.json` file if the version applies. This would be a bit opaque... but could also be an intended concequence of the changes, depending on how you look at it... I think I'm okay with this, but @octogonz should confirm that this is okay.

File: libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts (hunk 1 above)
Comment (D4N14L) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: I added my bits below. I think if we were to merge this, my only feedback at this point would be to update to use a proper `for` loop instead of the `forEach` method. We try to avoid doing that for readability reasons.

**Comment by octogonz** [comment]
@D4N14L a question we considered in Rush Hour: Should an implicitly preferred version be used to /narrow/ a preferred version? Or should an explicit preferred version always /replace/ it?

**Comment by D4N14L** [comment]
@octogonz I think it should _always_ replace it, as long as it is entirely a subset of the implicit version. This is the logic we follow in the pnpmfile shim, and this would make it consistent. If we want to limit it using the implicit version if it's a partial subset, that could be a future feature, but the base feature should align with what we already do IMO

File: libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts (hunk 1 above)
Comment (ace-n-msft) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: Actually - if we wanted to overwrite `implicitlyPreferredVersions` entirely, we could _probably_ drop this loop entirely.

Instead, we could simply swap lines 94 and 95 - so that the `implicitlyPreferredVersions` entries get overwritten during the second `mergeFromMap()` call.

Comment (D4N14L) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: @ace-n-msft I clarified below, but I think that we don't want to just blindly replace implicit preferred versions, but only replace if the explicit preferred version is entirely a subset of the implicit version, which matches existing behaviour in the pnpm shim

Comment (ace-n-msft) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: @D4N14L making sure I understand your viewpoint: you're saying we should **merge this PR as-is**, correct?

Comment (D4N14L) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: @ace-n-msft as-is, + use a proper for loop :)

Comment (ace-n-msft) [reply libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:101]: @D4N14L done - apologies for the delay!

**Review by D4N14L — APPROVED** [review]
